### PR TITLE
SAK-47290 Add ability to create criterion groups for rubrics

### PR DESCRIPTION
**Pull request by stetsche**
Demo [on the jira](https://sakaiproject.atlassian.net/browse/SAK-47290)
original commit from UPV by @josecebe

### CODE REVIEW

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js [hunk 1]
```diff
@@ -170,6 +187,10 @@ export class SakaiRubricCriteria extends RubricsElement {
           <span tabindex="0" role="button" class="add fa fa-plus"></span>
           <sr-lang key="add_criterion">Add Criterion</sr-lang>
         </button>
+        <button class="add-empty-criterion" @click="${(event) => this.createCriterion(event, true)}">
+          <span tabindex="0" role="button" class="add fa fa-plus"></span>
```
[the remark this answers is not in the document]
Comment (stetsche) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js:191]: Looks like false positive

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js [hunk 2]
```diff
@@ -24,51 +24,64 @@ export class SakaiRubricCriteriaReadonly extends RubricsElement {
...
-                              (${parseFloat((r.points * (c.weight / 100)).toFixed(2)).toLocaleString(this.locale)})
-                            </b>`
-                          : ""
-                        }
-                        ${r.points.toLocaleString(this.locale)}
-                        <sr-lang key="points">Points</sr-lang>
-                      </span>
+            ` : html`
+              <div id="criterion_row_${c.id}" class="criterion-row">
+                <div tabindex="0" class="criterion-detail">
```
Comment (ottenhoff) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js:36]: why is this div tabbable?

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ottenhoff) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js:191]: It looks like an ok comment to me.

* Why is the empty icon tabbable? What would a keyboard user do with it?
* What does it have a button role if clicking does nothing?

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-grading.js [hunk 3]
```diff
@@ -92,79 +92,88 @@ export class SakaiRubricGrading extends RubricsElement {
...
+                    </div>`
+                  : ""
+                }
+              </div>
+              <div class="criterion-ratings">
+                <div class="cr-table">
+                  <div class="cr-table-row">
+                  ${c.ratings.map(r => html`
+                    <div class="rating-item ${r.selected ? "selected" : ""}"
+                          tabindex="0"
```
[the remark this answers is not in the document]
Comment (stetsche) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-grading.js:125]: I agree, and I am not sure why it's on that div. But my edit here is just indenting that block by 4 spaces. As far as I am awar I am not introducing any new a11y related patterns here. I totally agree accessibility issues should be addressed, but I think it does not fit into this Jira ticket. What do you think?

[the remark this answers is not in the document]
Comment (ottenhoff) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-grading.js:125]: yes, agreed. that's the curse of changing indentation!

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (stetsche) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js:191]: I called this a false positive because on line 192 there is `<sr-lang key="save_weights">Save Weights</sr-lang>`. Which I assumed addresses this. But the tabindex seems out of lace, especially within a button :thinking:...  Again, this is exactly the same pattern, as with the button next to it (the lines above)

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-grading.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpellicer) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-grading.js:125]: Thanks for the review Sam.

Markus, fix this A11y issue in this PR please, or tell @V100X to do it!

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (stetsche) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js:191]: we will take a look

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java [hunk 4]
```diff
@@ -366,6 +366,29 @@ public Optional<CriterionTransferBean> createDefaultCriterion(String siteId, Lon
         });
     }
 
+    public Optional<CriterionTransferBean> createDefaultEmptyCriterion(String siteId, Long rubricId) {
+
+        String currentUserId = sessionManager.getCurrentSessionUserId();
+
+        if (StringUtils.isBlank(currentUserId) || !isEditor(siteId)) {
+            throw new SecurityException("You must be a rubrics editor to create/edit criteria");
+        }
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java:375]:
```suggestion

        if (!isEditor(siteId)) {
            throw new SecurityException("You must be a rubrics editor to create/edit criteria");
        }
```

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java [hunk 5]
```diff
@@ -366,6 +366,29 @@ public Optional<CriterionTransferBean> createDefaultCriterion(String siteId, Lon
...
+
+        return rubricRepository.findById(rubricId).map(rubric -> {
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java:377]: I'd get the rubric with an orElseThrow and an illegalargumentexception

Rubric rubric = rubricRepository.findById(rubricId).orElseThrow(() -> new IllegalArgumentException("No rubric for id " + rubricId));

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java [hunk 6]
```diff
@@ -366,6 +366,29 @@ public Optional<CriterionTransferBean> createDefaultCriterion(String siteId, Lon
...
+
+            Criterion criterion = new Criterion();
+            criterion.setOwnerId(siteId);
+            criterion.setTitle(resourceLoader.getString("default_empty_criterion_title"));
+
+            //criterion.setRubric(rubric);
```
Comment (adrianfish) [inline rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java:383]: Lose this :)

File: rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java (hunk 5 above)
Comment (adrianfish) [reply rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java:377]: I know the map stuff's really cool, but it makes sense to me to add some handling for a bad rubricId.

Comment (adrianfish) [reply rubrics/impl/src/main/java/org/sakaiproject/rubrics/impl/RubricsServiceImpl.java:377]: But, it's up to you

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/RubricsRestController.java [hunk 7]
```diff
@@ -175,6 +175,16 @@ ResponseEntity<EntityModel<CriterionTransferBean>> getDefaultCriterion(@PathVari
             .orElseGet(() -> ResponseEntity.badRequest().build());
     }
 
+    @GetMapping(value = "/sites/{siteId}/rubrics/{rubricId}/criteria/defaultEmpty", produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/RubricsRestController.java:178]: I feel like this should be a PUT. In effect you're putting a new criterion on the rubric. Maybe the url path should be /sites/{siteId}/rubrics/{rubricId}/criteria/empty

File: webapi/src/main/java/org/sakaiproject/webapi/controllers/RubricsRestController.java [hunk 8]
```diff
@@ -175,6 +175,16 @@ ResponseEntity<EntityModel<CriterionTransferBean>> getDefaultCriterion(@PathVari
...
+    ResponseEntity<EntityModel<CriterionTransferBean>> getDefaultEmptyCriterion(@PathVariable String siteId, @PathVariable Long rubricId) {
+
+        checkSakaiSession();
+
+        return rubricsService.createDefaultEmptyCriterion(siteId, rubricId)
+            .map(criterion -> ResponseEntity.ok(entityModelForCriterionBean(criterion)))
+            .orElseGet(() -> ResponseEntity.badRequest().build());
```
Comment (adrianfish) [inline webapi/src/main/java/org/sakaiproject/webapi/controllers/RubricsRestController.java:185]: I like this. This is nice.

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js [hunk 9]
```diff
@@ -24,51 +24,64 @@ export class SakaiRubricCriteriaReadonly extends RubricsElement {
...
-                            <b>
-                              (${parseFloat((r.points * (c.weight / 100)).toFixed(2)).toLocaleString(this.locale)})
-                            </b>`
-                          : ""
-                        }
-                        ${r.points.toLocaleString(this.locale)}
-                        <sr-lang key="points">Points</sr-lang>
-                      </span>
+            ` : html`
+              <div id="criterion_row_${c.id}" class="criterion-row">
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js:35]: Can we stick to the pattern of hyphenate4d ids and  classes?

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js [hunk 10]
```diff
@@ -24,51 +24,64 @@ export class SakaiRubricCriteriaReadonly extends RubricsElement {
...
+                      <div class="criterion-weight">
+                          <span>
+                            <sr-lang key="weight">Weight</sr-lang>
+                          </span>
+                          <span>${c.weight.toLocaleString(this.locale)}</span>
+                          <span>
+                            <sr-lang key="percent_sign">%</sr-lang>
+                          </span>
+                      </div>`
+                    : ""
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js:49]: Can we stick to 2 spaces for tabs

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js [hunk 11]
```diff
@@ -24,51 +24,64 @@ export class SakaiRubricCriteriaReadonly extends RubricsElement {
...
-            </div>
+            `}
           `)}
         </div>
       </div>
     `;
   }
+
+  isCriterionGroup(criterion) {
+    return criterion.ratings.length == 0;
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js]: === (pedantic, I know).

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js [hunk 12]
```diff
@@ -65,89 +65,106 @@ export class SakaiRubricCriteria extends RubricsElement {
...
+                    site-id="${this.siteId}"
+                    rubric-id="${this.rubricId}"
+                    criterion="${JSON.stringify(c)}">
+                </sakai-rubric-criterion-edit>
+              </h4>
+              <p>
+                ${unsafeHTML(c.description)}
+              </p>
+              ${this.weighted ? html`
+                  <div class="form-inline weight-field">
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js:99]: 2 spaces for tabs, please :)

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js [hunk 13]
```diff
@@ -457,6 +476,14 @@ export class SakaiRubricCriteria extends RubricsElement {
       this.cloneCriterion(e);
     }
   }
+
+  isRatingRemovable(criterion) {
+    return criterion.ratings.length > 1;
+  }
+
+  isCriterionGroup(criterion) {
+    return criterion.ratings.length == 0;
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria.js]: ===

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js [hunk 14]
```diff
@@ -16,52 +16,65 @@ export class SakaiRubricCriterionPreview extends RubricsElement {
...
-                    <h5 class="criterion-item-title">${r.title}</h5>
-                    <div class="div-description">
-                      <p>${r.description}</p>
+          ` : html`
+            <div id="criterion_row_${c.id}" class="criterion-row">
+              <div class="criterion-detail">
+                <h4 class="criterion-title">${c.title}</h4>
+                <p>${unsafeHTML(c.description)}</p>
+                ${this.weighted ? html`
+                    <div class="criterion-weight">
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js:32]: 2 spaces for tabs, please

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js [hunk 15]
```diff
@@ -16,52 +16,65 @@ export class SakaiRubricCriterionPreview extends RubricsElement {
...
+                  <div class="cr-table-row">
+                  ${c.ratings.map(r => html`
+                    <div class="rating-item" id="rating_item_${r.id}" >
+                      <h5 class="criterion-item-title">${r.title}</h5>
+                      <div class="div-description">
+                        <p>${r.description}</p>
+                      </div>
+                      <span class="points">
+                        ${this.weighted && r.points > 0 ? html`
+                            <b>
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js:55]: I think you should have this in a span, with styles on it.

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js [hunk 16]
```diff
@@ -16,52 +16,65 @@ export class SakaiRubricCriterionPreview extends RubricsElement {
...
             </div>
-          </div>
+          `}
         `)}
       </div>
     `;
   }
+
+  isCriterionGroup(criterion) {
+    return criterion.ratings.length == 0;
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js]: ===

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-rating-edit.js [hunk 17]
```diff
@@ -10,6 +10,7 @@ export class SakaiRubricCriterionRatingEdit extends RubricsElement {
     return {
       rating: { type: Object },
       criterionId: { attribute: "criterion-id", type: String },
+      removable: {attribute: "removable", type: Boolean},
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-rating-edit.js:13]: You don't need the attribute for this, as it is the same.

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-rating-edit.js [hunk 18]
```diff
@@ -153,6 +154,13 @@ export class SakaiRubricCriterionRatingEdit extends RubricsElement {
       this.editRating(e);
     }
   }
+
+  removeButtonTitle() {
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-rating-edit.js:158]: I know some people find this pedantic and weird, but, please indulge and old man and leave a blank line after the function name.

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-rating-edit.js [hunk 19]
```diff
@@ -153,6 +154,13 @@ export class SakaiRubricCriterionRatingEdit extends RubricsElement {
...
+    if (this.removable) {
+      return `${tr("remove_label")} ${this.rating.title}`;
+    }
+    return tr("remove_rating_disabled");
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-rating-edit.js:162]:
```suggestion
    return this.removable ? `${tr("remove_label")} ${this.rating.title}` : tr("remove_rating_disabled");
```

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-student.js [hunk 20]
```diff
@@ -116,6 +123,10 @@ export class SakaiRubricCriterionStudent extends RubricsElement {
     `;
   }
 
+  isCriterionGroup(criterion) {
+    return criterion.ratings.length == 0;
```
Comment (adrianfish) [inline webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-student.js]: ===

**Review by adrianfish — CHANGES_REQUESTED** [review]
Mainly small stuff. Generally, this is a nicely done change. Thanks.

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js (hunk 16 above)
Comment (stetsche) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criterion-preview.js]: > Mainly small stuff. Generally, this is a nicely done change. Thanks.

Generally, when I am working on a feature like this I am trying to stick to present patterns in the code. Most of what you commented on is STRONGLY inspired (pretty much a copy), by the existing code, which seems to be authored by you. I think it is fine if you change your mind about how that should be written. But let's please not rethink everything once we want to add a feature. I think this changes should be seperate, and are not related to the feature.

TLDR: Let's focus on what functionality is added, not on which patterns are inherited from the existing codebase.

**Comment by stetsche** [comment]
> Mainly small stuff. Generally, this is a nicely done change. Thanks.

Generally, when I am working on a feature like this I am trying to stick to present patterns in the code. Most of what you commented on is STRONGLY inspired (pretty much a copy), by the existing code, which seems to be authored by you. I think it is fine if you change your mind about how that should be written. But let's please not rethink everything once we want to add a feature. I think this changes should be seperate, and are not related to the feature.

TLDR: Let's focus on what functionality is added, not on which patterns are inherited from the existing codebase.

**Review by mpellicer — CHANGES_REQUESTED** [review]
I tested this locally and found some bugs in the preview and the group description, the preview displays the wrong order of the groups, and the group description is disabled when editing a group description. Could you fix these issues? I'll retest.

Thanks

**Comment by adrianfish** [comment]
@stetsche Fair enough on the formatting. I've left those 4 space tab sections in by mistake and would have changed them myself if I was in the code, that's all.

File: webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js (hunk 2 above)
Comment (stetsche) [reply webcomponents/tool/src/main/frontend/js/rubrics/sakai-rubric-criteria-readonly.js:36]: @adrianfish could you review, why this is focusable please? I'm not sure whats the purpose here

**Review by mpellicer — APPROVED** [review]
After more testing the group description issue is not related to this and the preview works fine, the PDF problem depends on a different issue, thanks for the good work @josecebe and @stetsche !

**Comment by mpellicer** [comment]
Thanks for also fixing the codacy issues.

**Comment by ottenhoff** [comment]
Hi @mpellicer , it feels like this was merged while conversations were still happening .... I raised an accessibility concern that I feel like we should have ironed out.

**Comment by mpellicer** [comment]
Hey Sam, 

I told the team to fix it:
image

and Markus is aware of the bug:
image

We agreed to fix the accessibility bug in a separate PR since it's not caused by these changes. Does that make sense? He just changed the identation,  the bug will be resolved using a different SAK-, probably @V100X will work on it since he's dealing with the A11y stuff.

**Comment by mpellicer** [comment]
> Hi @mpellicer , it feels like this was merged while conversations were still happening .... I raised an accessibility concern that I feel like we should have ironed out.

To be fair, this is not the only PR that is being merged when conversations are still happening, and they are merged without a single explanation, but that's other issue and maybe this is not the right place to talk about it!

**Comment by adrianfish** [comment]
I'm getting a securityexception when creating a group criteria

**Comment by adrianfish** [comment]
It could be my local code. I'm just double checking.

**Comment by adrianfish** [comment]
No, not my code. I think somebody else needs to test this.

**Comment by stetsche** [comment]
@adrianfish Yes, I could reproduce it too as an instructor. Thanks for offering to work on it on [SAK-47315](https://sakaiproject.atlassian.net/browse/SAK-47315)!
@ottenhoff Yes, we are addressing this a11y issues ;)

**Comment by mpellicer** [comment]
> @adrianfish Yes, I could reproduce it too as an instructor. Thanks for offering to work on it on [SAK-47315](https://sakaiproject.atlassian.net/browse/SAK-47315)! @ottenhoff Yes, we are addressing this a11y issues ;)

If we introduced SAK-47315 in this piece of work please fix it @stetsche, we always take responsability of our contributions. Same for any new a11y issue, if the problem already existed maybe assign it to @V100X since he's dealing with many existing a11y issues. Thanks.

**Comment by adrianfish** [comment]
No, Markus didn't introduce it, it's a remnant of the refactor I think. And, the a11y stuff has been there since the rubrics tool was originally written (not by by me, either). Let me get my fix to SAK-47315 in and we can look at the divs.
